Improve Error Reporting of SemVer Parser

vinayman opened the pull request.

chriseth committed:
Unused store eliminator.

whhygee committed:
CI: shared build and dependency installation steps

whhygee committed:
Ci: parameterize the clang sanitizer jobs

bshastry committed:
Yul interpreter: Add flag to disable memory tracing and dump for fuzzing.
Model revert in yul interpreter.
Add logTrace for a few more instructions and clear trace on revert.

nikitastupin committed:
Separate visibility for state variables and functions

hrkrshnn committed:
Extend using-for.

cameel committed:
Update gnosis external test to use upstream directly and use Hardhat

cameel committed:
externalTests: Add support for hardhat.config.ts

chriseth committed:
Extract external function part into its own function.

ekpyron committed:
Fix util::valueOrDefault.

ekpyron committed:
Build packed soljson.js.

joshuatarkwski committed:
Output searched locations on import failure.

chriseth committed:
Equal store eliminator.

chriseth committed:
Test updates.

chriseth committed:
Using for with global binding.

chriseth committed:
Documentation

cameel committed:
External test for sushiswap/trident

cameel committed:
trident ext test: Patch upstream instread of relying on our fork

caffeinum committed:
docs: note on invalid SPDX `UNLICENSED`

https://github.com/ethereum/solidity/issues/12251

nishant-sachdeva committed:
Equality operator allowed for external function types

cameel committed:
externalTests: Preset selection via command-line arguments

cameel committed:
Python script for parsing eth-gas-reporter output

cameel committed:
externalTests: Clean the build/ dir for Hardhat too

cameel committed:
externalTests: Benchmark reports

chriseth committed:
Use options struct for function type factory function.

chriseth committed:
Add option member for function type.

chriseth committed:
Use function names instead of members.

christianparpart committed:
LSP: Implements goto-definition.

cameel committed:
CI: Create templates for parameterized jobs to make it easier to reuse them in multiple workflows

cameel committed:
externalTests: Make comments about failing presets less terse

cameel committed:
CI: Add missing workflow_trigger_on_tags filter to bytecode jobs

cameel committed:
Benchmark report collector job + summary

semaraugusto committed:
fix issue with delegating votes to people who cannot vote. That made it so that if wallet A delegated to wallet B and B had no rights to vote, than B would lose out on its own vote

semaraugusto committed:
Update docs/examples/voting.rst

Co-authored-by: Bhargava Shastry <bshastry@ethereum.org>

chriseth committed:
Fix mapping example.

Marenz committed:
Properly resolve virtual modifiers

hrkrshnn committed:
Added more tests. Updated chromosome. Remove unused headers.

christianparpart committed:
LSP: Introduce HandlerError(id, code, message) exception for easier handling.

ekpyron committed:
Use fixed seed for flaky phaser test for now.

fulldecent committed:
Typo: SemVer

christianparpart committed:
LSP: Introduces lspAssert(condition, ErrorCode, message)

christianparpart committed:
LSP: Introduce HandlerBase for future LSP-feature implementations.

esquith committed:
fix typo

bshastry committed:
Add FunctionHoister as a pre-requisite for equal store eliminator.

safwen1 committed:
Explaining payable(msg.sender)

At this point of the documentation a new Solidity learner will not understand this line without further explanation:
if (!payable(msg.sender).send(amount)) {
It should explain how msg.sender is of type "address" and not "address payable" so it cannot send or receive Ether. Therefore it must be explicitly converted to payable.

Commit:
Fixing typo in "smart contracts"

cameel committed:
External test for ElementFi

cameel committed:
Include the parallel run number in gitter notifcations from CI

cameel committed:
externalTests: Add support for cloning repos at a specific commit

cameel committed:
Switch the trident external test to and earlier, working revision

christianparpart committed:
Adds missing accepts to ImportDirective's SymbolAlias::symbol

cameel committed:
trident: Sort test suites to make the order the same in CI and locally

cameel committed:
trident: Note explaining the hard-coded commit

cameel committed:
trident: Update to match changes done to other external tests in the preset selection PR

ekpyron committed:
Update docker images to Z3 4.8.14.

ekpyron committed:
Test case for current behaviour.

Marenz committed:
Fix ICE when a constant variable declaration forward references a struct

Commit:
update smtchecker tests for new z3

Marenz committed:
Fix genetic algorithms CI failure

ekpyron committed:
Update osx z3 archive hash.

ekpyron committed:
Remove counterexample from test.

Commit:
adjust tests for nondeterminism

Marenz committed:
Fix wrong error with immutables when base contract c'tor uses return

fulldecent committed:
Document address members code and codehash

ekpyron committed:
Adjust documentation for linux packages not maintained by us.

cameel committed:
externalTests: Add force_hardhat_unlimited_contract_size

cameel committed:
External test for Euler

cameel committed:
ens: Remove global solc replacement (only necessary with Truffle)

cameel committed:
External test for Yield Liquidator V2

cameel committed:
External test for Bleeps

cameel committed:
externalTests: Add name_hardhat_default_export

cameel committed:
External test for Uniswap V3

ekpyron committed:
Fill in junk in stack layouts on terminating control flow paths.

cameel committed:
CI: Use 'large' resource class for macOS and use -j without spaces

cameel committed:
External test for PRBMath

Marenz committed:
Natspec: Fix ICE when overriding a struct getter with a Natspec-documented return value and the name in the struct is different.

ekpyron committed:
Allow builtins as yul identifier paths in antlr grammar.

ekpyron committed:
Update test/libsolidity/syntaxTests/inlineAssembly/assignment_to_function_pointer.sol

Co-authored-by: Kamil Śliwak <kamil.sliwak@codepoets.it>

joshieDo committed:
Add event and error identifiers to cli hashes cmd

Marenz committed:
Add test for fixed natspec ICE

Marenz committed:
Fix .push() not considering external functions

cameel committed:
CI: Remove the compile-only run for Gnosis

Marenz committed:
Require strict parsing for natspec JSON tests

Marenz committed:
Make formatting more consistent in natspec tests

nishant-sachdeva committed:
Adding Stack Height Checker and modifying the number of POP instructions to appropriately provide the pointer address

cameel committed:
External tests for Pool Together v4

cameel committed:
force_hardhat_unlimited_contract_size: Initialize 'hardhat' network if is not explicitly configured

cameel committed:
External test for Tracer Perpetual Pools

Marenz committed:
Fix too strict assert for calldata string -> bytes conversions

Marenz committed:
Remove empty assert messages

brunobar79 committed:
Add modifiers to the order of layout

ekpyron committed:
Improved stack shuffling in corner cases.

cameel committed:
externalTests: Make force_hardhat_unlimited_contract_size work with Typescript < 4.0

- The `??=` operator was added in Typescript 4. `??` Was already available on 3.7 so we can use it instead to get projects like Uniswap work.

gorgos committed:
Add msg.sig to cheatsheet

hrkrshnn committed:
Fixed a ICE on calldata to struct member copy

cameel committed:
CI: Fix job name for PRBMath external test

cameel committed:
Change the way we run solcjs now that it has been renamed in solc-js

cameel committed:
bytecode compare: Stop depending on solc-js internal structure and install it via npm from a local dir instead

cameel committed:
Changelog entry for the TypeScript migration

nishant-sachdeva committed:
Corresponding code in the .cpp file has been commented instead of begin removed pending preliminary reviews
Code generators needed fixing of the cleanup process during typecasting of bytes and integers

nishant-sachdeva committed:
Corresponding code in the .cpp file has been commented instead of begin removed pending preliminary reviews
Code generators needed fixing of the cleanup process during typecasting of bytes and integers

cameel committed:
Release checklist: switch to a command that automatically downloads the binary before running verifyVersion

cameel committed:
Adjust workflow with solc-js checkout now that it's on TypeScrpt

cameel committed:
fixup! bytecode compare: Stop depending on solc-js internal structure and install it via npm from a local dir instead

fulldecent committed:
Document our version policy w/r/t SemVer

bshastry committed:
Filter log statements and remove gas comparison.

Marenz committed:
Emit immutable references for pure yul code

bshastry committed:
Code transform fuzzer: Only test against latest EVM version.

Younghoon-Lee committed:
Add blank line to make it consistency

nishant-sachdeva committed:
Changed occurences of isByteArray() to isByteArrayOrString(). The idea
is to, in a future commit, replace such occurences of
isByteArrayOrString() which are required to return True only for Bytes
type with a new isByteArray() function.

cameel committed:
Make solc-js.sh run its tests directly

- Its structure has diverged a lot from other external tests and there's not point in keeping it abstracted like this.

cameel committed:
externalTests/solc-js: Allow using a local checkout of solc-js

cameel committed:
Update the release checklist to use the build+publish commands for solc-js

0xsomnus committed:
Added solidity visual auditor for VSCode

franzihei committed:
Adding a few resources

Marenz committed:
Merge identifier query methods into one

chriseth commented on liblangutil/SemVerHandler.cpp:
```diff
@@ -265,14 +275,30 @@ unsigned SemVerMatchExpressionParser::parseVersionPart()
+			{
+				BOOST_THROW_EXCEPTION(
+					SemVerError(
+						1465_error,
```
You cannot re-use the same error id in multiple locations, just create new (random) error IDs.

chriseth commented on liblangutil/SemVerHandler.cpp:
```diff
@@ -171,12 +171,22 @@ optional<SemVerMatchExpression> SemVerMatchExpressionParser::parse()
-	catch (SemVerError const&)
+	catch (SemVerError const& e)
 	{
+		BOOST_THROW_EXCEPTION(e);
```
I don't think it is good to throw an exception here. Instead, you could return an `std::variant<SemVerError, SemVerMatchExpression>` here and handle it accordingly in the parser to create the error.

chriseth committed:
Correct type of address.code

chriseth committed:
Fix changelog

cameel committed:
perpetual-pools: Switch the test to our fork (original repo is gone)

cameel committed:
Add missing SELECTED_PRESETS argument to some of the recently added external tests

Marenz committed:
Include used events in ``--hashes`` output

cameel committed:
bleeps: Switch to the main branch

cameel committed:
bleeps: Patch WETH9.sol for 0.8.x

cameel committed:
Disable bleeps external test until it gets fixed upstream

nvnx7 committed:
plain `address` can be sent Ether too

The docs state that a plain `address` cannot be sent Ether. But even though `send` and `transfer` members are not available for plain `address`, the `call` is. And `call` can be invoked upon a plain `address` type to send Ether to the address.

For instance, the `someone` (`address` type) can be sent Ether by invoking `sendSomeone()` method in the following `Dummy` contract:

```
contract Dummy {
    address someone = 0xAb8...cb2;

    function balanceOf(address addr) public view returns (uint) {
        return addr.balance;
    }

    function sendToSomeone() public payable returns (bool) {
        (bool sent, ) = someone.call{value: msg.value}("");
        return sent;
    }
}
```

nishant-sachdeva committed:
Changed instaces of isByteArrayOrString() to isByteArray() where it's only supposed to return a True for Bytes Type

ekpyron committed:
Consider all grandparents in override analysis.

fulldecent committed:
Add NatSpec note for libraries

ekpyron committed:
Allow annotating inline assembly as memory-safe.

ekpyron committed:
Patch external tests with a safe inline assembly annotation.

ekpyron committed:
Report memoryguard in stack too deep error.

cameel committed:
When installing solc-js use the dist/ subdir, which contains the built JS files

shuklaayush committed:
Fix slot calculation for bytes/string mapping

ekpyron committed:
Documentation.

nishant-sachdeva committed:
Types.h:872 had a comment /// String is interpreted as a subtype of
Bytes. - this was now incorrect after #12593 . That has been removed
now.

nishant-sachdeva committed:
Added support for FunctionType::Kind::StringConcat and functions string.concat

cameel committed:
Re-enable Bleeps and just disable the failing governor test

cameel committed:
Run full tests, not just test:contracts in PRBMath external test

- `test:contracts` does not seem to be running any tests at all.

a3d4 committed:
Clarify symlink handling on Windows

cameel committed:
CI: Rename node_latest_small to node_small

cameel committed:
CI: Switch t_ems_ext_hardhat to nodejs 16

cameel committed:
CI: Remove notes about Hardhat failing on nodejs 17; that version is officially not supported

a3d4 committed:
Treat root path in normalizeCLIPathForVFS as case insensitive on Windows

a3d4 committed:
Remove a reference to unused SolidityFixedFeeRegistrar

fulldecent committed:
Fix signature of pop member

Hakeemmidan committed:
Update cheatsheet.rst

Add more description to the bullet point of 'block.timestamp' (under 'Global Variables')

chriseth committed:
Re-enable preset for poolTogether.

a3d4 committed:
Enable asserts in MSVC RelWithDebInfo configuration

ekpyron committed:
Add emscripten's mini-lz4.js.

ekpyron committed:
Adjust mini-lz4.js and base64DecToArr.js.

ekpyron committed:
Add base64DecToArr.js.

cameel committed:
CI: Store all artifacts in steps_build so that we can use it for b_ubu

fulldecent committed:
Note about storage compatability mode

ekpyron committed:
Move stray semantics test to proper directory.

nishant-sachdeva committed:
Added errorSelector field for custom errors

cameel committed:
storebytecode.sh: Update outdated script docstring.

ekpyron committed:
Inline assembly without memory effects is implicitly memory safe.

ekpyron committed:
Tests.

cameel committed:
wasm-rebuild: Incorrect comment about script location

paarthmadan committed:
Add Base64 test cases with and without inline assembly

Adds two implementations of Base64 encoding as specified in RFC4648.
Implementation (1) uses inline assembly, while Implementation (2) is
written purely in Solidity.

Assertions are added to replicate the test vectors specified in the RFC
for Base64 to ensure both implementations to specification.

amsavarthan committed:
Highlighted "revert" command

When reading through the documentation, I found that highlighting was missed for the revert command and added that.

ldfp committed:
Solve issue 12658 ; Solving wrong syntax.

Solve this issue : https://github.com/ethereum/solidity/issues/12658

ekpyron committed:
Disable running smt tests in ubsan runs.

ekpyron committed:
Tests.

ekpyron committed:
Allow memory-safe inline assembly dialect flag.

ekpyron committed:
Update docs.

ekpyron committed:
Adjust grammar.

chriseth committed:
Minor changes to changelog.

ekpyron committed:
Add Changelog entry about packed soljson.js.

Commit:
Add comment

chriseth committed:
Update tests.

cameel committed:
elementfi: Bypass expiration check that won't pass due to test settings

chriseth committed:
Extend side effects to operations.

fulldecent committed:
Fix typos

devtooligan committed:
fix: corrects _ prefixes

devtooligan committed:
fix: remove lower_case_with_underscores and Capitalized_Words_With_Underscores from code examples

chriseth committed:
Set release date.

nishant-sachdeva committed:
returnSize assigned according to truth value of returnInfo.dynamicReturnSize

cameel committed:
Neutralize packaged Uniswap contracts in yield-liquidator external test

chriseth committed:
Set version to 0.8.13

callezenwaka committed:
Update typo in using-for.rst

Update typo in using-for.rst

chriseth committed:
Update PPA build script.

chriseth committed:
Add codeowners file.

wesbz committed:
Fixing a typo in doc's EVM dialect in Yul's doc.

It's the tiniest typo but it's visible on the website.

Commit:
Add not_existing_import.json to ASTJSON tests

Commit:
Fix syntax errors in ASTJSON tests

Marenz committed:
Make error message less confusing

cameel committed:
wasm-rebuild: Revert solc-js changes - not needed because the script uses an old branch of solc-js

cameel committed:
wasm-rebuild: Upgrade packages to get newer SSL certificates required to be able to download boost over HTTPS

cameel committed:
wasm-rebuild: Patch storebytecode.sh from 0.6.1 not to use latest solc-js

Commit:
Non-fatal erros do not stop ASTJSON test execution

cameel committed:
wasm-rebuild: Print tags that will be processed

cameel committed:
wasm-rebuild: Update solc-bin URL

Abdulkarim4u committed:
fixed grammar on doc.

If readers can gain additional pieces of  information on smart contracts, then  "more details" should be correct.

a3d4 committed:
Fix running tests from any drive

aarlt committed:
scripts/common.sh: 'diff --color' not supported by macOS.

chriseth committed:
Update .github/CODEOWNERS

Co-authored-by: Kamil Śliwak <kamil.sliwak@codepoets.it>

Commit:
Support compilation fail scenarios in ASTJSON tests

nishant-sachdeva committed:
Generate warning when rational numbers are converted to their mobile type without explicit requests

Commit:
Add "fail after parsing" scenario to ASTJSON tests

Commit:
Generate ASTJSONTest variants only if a file with expected result exists

cameel committed:
Remove release.sh/.bat scripts we no longer use

cameel committed:
Add instructions for creating the source tarball back to the release checklist

Commit:
Use "C" instead user environment locale in solc

cameel committed:
A script and CI job for testing the compiler with different locales

cameel committed:
Disable flaky ElementFi tests

chriseth committed:
Fix iterable mapping example.

Commit:
Ignore cex in SMT test

franzihei committed:
updates index with language links

cameel committed:
Re-enable running tests in external tests via IR in cases where they don't pass due to Hardhat heuristics

aarlt committed:
Update buildpacks to include jq.

aarlt committed:
Update CI to use new buildpacks.

chriseth committed:
Split out disassemble to remove numeric from instruction includes.

chriseth committed:
Also save and restore the nesting depth.

cameel committed:
perpetual-pools: Use Hardhat 2.8.4 to work around @openzeppelin/hardhat-upgrades's problem with parallel compilation on Hardhat 2.9.0

wechman committed:
Replace all locale-dependent operations with locale-agnostic counterparts

david-k committed:
Fix leftover use of divModWithSlacks in doc

The option `divModWithSlacks` was previously changed to `divModNoSlacks`.
However, this was not reflected in the documentation.

msooseth committed:
Fixing style issue in ExecutionFramework.cpp

It should be `auto const&` not `const auto&`

tylerfxn committed:
Remove use of `using namespace` in header file

tylerfxn committed:
Specify namespaces

Fix references into solidity::util

ekpyron committed:
Remove unused entry point to evmasm Assembly.

ekpyron committed:
Store whether an evmasm Assembly is creation code.

ekpyron committed:
Add test case and adjust tests.

christianparpart committed:
[CMake] Refactor libsolc linker flag for EXPORTED_FUNCTIONS and make it more readable.

wechman committed:
Use sol file to validate test variants

ekpyron committed:
Adjust documentation.

ekpyron committed:
Remove isCreation flag from OptimiserSettings.

chriseth committed:
Optimize eq iszero jumpi to xor jumpi and remove double jump.

chriseth committed:
Provide generic access to type definition for user-defined types.

ekpyron committed:
Mark viaIR code generation as non-experimental.

chriseth committed:
Simplify peephole optimizer template.

ekpyron committed:
Unit test.

Commit:
doc: improve explanation of interfaces

cameel committed:
euler: Override Hardhat test timeout

chriseth committed:
Remove unneeded header files.

cameel committed:
euler: Use project's own TEST_TIMEOUT var for changing the timeout instead of mocha settings

chriseth committed:
Refactor data flow analyzer state access.

ekpyron committed:
Test for old behaviour.

ekpyron committed:
Fix and updated test.

chriseth committed:
Create a state struct.

tylerfxn committed:
more qualifying

tylerfxn committed:
Fix CommonSyntaxTest.cpp and others

nishant-sachdeva committed:
Error messages from abi encodecall are giving more details about the types of the involved variables

ekpyron committed:
Fix import directive visits in type checker and view pure checker.

ekpyron committed:
Change memory-safe assembly annotation suggestion during stack too deep errors.

chriseth committed:
Clean up changelog.

bshastry committed:
Permit control flow mutations in global scope of a Yul program.

ekpyron committed:
No longer mark Yul as experimental.

chriseth committed:
Using for at file level.

chriseth committed:
Add helper to see if a switch has a default case.
